discovery_client.go

```diff
@@ -0,0 +1,134 @@
+package rendezvous
+
+import (
+	"context"
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"math"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+type rendezvousDiscoveryClient struct {
+	rp        RendezvousPoint
+	peerCache sync.Map //is a map[string]discoveredPeerCache
+	rng       *rand.Rand
+	rngMux    sync.Mutex
+}
+
+type discoveredPeerCache struct {
+	cachedRegs []Registration
+	mux        sync.Mutex
+}
+
+func NewRendezvousDiscoveryClient(host host.Host, rendezvousPeer peer.ID) discovery.Discovery {
+	rp := NewRendezvousPoint(host, rendezvousPeer)
+	return &rendezvousDiscoveryClient{rp, sync.Map{}, rand.New(rand.NewSource(rand.Int63())), sync.Mutex{}}
+}
+
+func (c *rendezvousDiscoveryClient) Advertise(ctx context.Context, ns string, opts ...discovery.Option) (time.Duration, error) {
+	// Get options
+	var options discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return 0, err
+	}
+
+	ttl := options.Ttl
+	var ttlSeconds int
+
+	// Default is minimum duration
+	if ttl == 0 {
+		ttlSeconds = 120
+	} else {
+		ttlSeconds = int(math.Round(ttl.Seconds()))
+	}
+
+	if err := c.rp.Register(ctx, ns, ttlSeconds); err != nil {
+		return 0, err
+	}
+
+	actualTTL := time.Duration(ttlSeconds) * time.Second
+	return actualTTL, nil
+}
+
+func (c *rendezvousDiscoveryClient) FindPeers(ctx context.Context, ns string, opts ...discovery.Option) (<-chan peer.AddrInfo, error) {
+	// Get options
+	var options discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	const maxLimit = 1000
+	limit := options.Limit
+	if limit == 0 || limit > maxLimit {
+		limit = maxLimit
+	}
+
+	// Get cached peers
+	var cache *discoveredPeerCache
+
+	genericCache, _ := c.peerCache.LoadOrStore(ns, &discoveredPeerCache{})
+	cache = genericCache.(*discoveredPeerCache)
+
+	cache.mux.Lock()
+	cachedRegs := cache.cachedRegs
+
+	// Remove all expired entries from cache
+	currentTime := int(time.Now().Unix())
+	newCacheSize := len(cachedRegs)
+
+	for i := 0; i < newCacheSize; i++ {
+		reg := cachedRegs[i]
+		if reg.Ttl < currentTime {
+			newCacheSize--
+			if i != newCacheSize {
+				cachedRegs[i] = cachedRegs[newCacheSize]
+				i--
+			}
+		}
+	}
+	cache.cachedRegs = cachedRegs[:newCacheSize]
+	cache.mux.Unlock()
+
+	// Discover new records if we don't have enough
+	var discoveryErr error
+	if newCacheSize < limit {
+		if discoveryRecords, _, err := c.rp.Discover(ctx, ns, limit, nil); err == nil {
```

Comment:
We should probably loop doing `rp.Discover` by using the cookie so that we get all peers.
This will stop just at the first set of peers and never discover any more.

Reply:
We talked about the max limit on this type of blocking client being about 1000 records which conveniently is the maximum limit for the rendezvous server. So one page should be enough until we do a non-blocking client.

Reply:
Let's still use the cookie to always make progress even if we only do a query at a time.

Reply:
To clarify, because of the internal caching utilizing the cookie will only help us make progress if we query beyond `limit - newCacheSize`. However, we can definitely use the cookie to help in this case. It should also serve to limit the data returned from repeated `limit=MaxLimit` queries.

Reply:
This should be implemented now. However, the way I randomly select items to return is not nearly as efficient as it could be. Given that this is just on the client side and the list sizes are small (<1000) it my not be necessary to get it fixed up now, however any suggestions are welcome.